drivers/stepper/gpio_stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,217 @@
+	int32_t target_position;
+	uint32_t delay_in_us;
+};
+
+static int32_t stepper_motor_move_with_4_gpios(const struct gpio_dt_spec *gpio_spec,
+					       int32_t step_number)
+{
+	switch (step_number) {
+	case 0: {
+		LOG_DBG("Step 1010 %d %d %d %d", gpio_spec[0].pin, gpio_spec[1].pin,
+			gpio_spec[2].pin, gpio_spec[3].pin);
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_INACTIVE);
+		break;
+	}
+	case 1: {
+		LOG_DBG("Step 0110");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_INACTIVE);
+		break;
+	}
+	case 2: {
+		LOG_DBG("Step 0101");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_ACTIVE);
+		break;
+	}
+	case 3: {
+		LOG_DBG("Step 1001");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_ACTIVE);
+		break;
+	}
+	}
+	return 0;
+}
+
+void stepper_work_step_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct gpio_stepper_motor_controller_data *data =
+		CONTAINER_OF(dwork, struct gpio_stepper_motor_controller_data, stepper_dwork);
+	const struct gpio_stepper_motor_controller_config *config = data->dev->config;
+
+	static int32_t step_count = 1;
+
+	K_SPINLOCK(&data->lock) {
+		LOG_DBG("Moving Motor from %d to %d with delay %d ", data->actual_position,
+			data->target_position, data->delay_in_us);
+		if (data->actual_position != data->target_position) {
+			if (data->actual_position > data->target_position) {
+				stepper_motor_move_with_4_gpios(config->control_pins, step_count--);
+				step_count =
+					step_count < 0 ? (NUMBER_OF_GPIO_PINS - 1) : step_count;
+				data->actual_position--;
+			} else if (data->actual_position < data->target_position) {
+				stepper_motor_move_with_4_gpios(config->control_pins, step_count++);
+				step_count =
+					step_count > (NUMBER_OF_GPIO_PINS - 1) ? 0 : step_count;
+				data->actual_position++;
+			}
+			k_work_reschedule(&data->stepper_dwork, K_USEC(data->delay_in_us));
+		}
+	}
+}
+
+static int32_t gpio_stepper_move(const struct device *dev, int32_t micro_steps)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->target_position = data->actual_position + micro_steps;
+		k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	}
+	return 0;
+}
```

Comment:
is move an async API? if so, why not allow clients to wait for completion?

Reply:
Atleast the way i've programmed it for now if async? Would you suggest to have both sync and async functions. With wait for completion you a mean to have a sync aka blocking call, or?

Reply:
done. Thanks to @faxe1008